rust/src/traits/samplers/bernoulli/mod.tex

```diff
@@ -0,0 +1,327 @@
+\end{proof}
+
+\subsection{Pseudocode \& proof of \texttt{sample\_geometric\_buffer}}
+\begin{lstlisting}[language=Python, escapechar=|]
+# returns a sample from the Geom(0.5) distribution
+def sample_geometric_buffer(B : uint, constant_time : bool) -> Optional[uint]: |\label{line:geombuffer}|
+    buf = bytearray(B)
+    buf = fill_bytes(buf)
+    
+    if (constant_time): 
+        ret = None
+        for i in range(len(buf)):
+            # find first nonzero event
+            if (buf[i] > 0):
+                # viewing buffer as string of bits, compute index of first nonzero bit
+                ret = 8 * i + leading_zeroes(buf[i]) |\label{line:indexcmp}|
+        return ret
+    else:
+        for i in range(len(buf)):
+            if (buf[i] > 0):
+                return(8 * i + leading_zeroes(buf[i]))
+        
+        return None
+\end{lstlisting}
+
+\begin{tcolorbox}
+\begin{theorem}[\texttt{sample\_geometric\_buffer}]
+The function \texttt{sample\_geometric\_buffer(B, constant\_time)} returns a random variable $X \sim CondGeom(0.5)$, a draw from a conditional Geometric distribution conditioned on the event that $Geom(0.5) < 8 \cdot B$. Otherwise \texttt{sample\_geometric\_buffer} returns \texttt{None}. 
+\end{theorem}
+\end{tcolorbox}
+
+\begin{proof}
+The function \texttt{sample\_geometric\_buffer} uses \texttt{fill\_bytes}  as a subroutine to generate a buffer of $B$ bytes where for each bit $b$ in the buffer it follows that $\Pr[b = 1] = \frac{1}{2}$ and $\Pr[b = 0] = \frac{1}{2}$ by Lemma 2.2. Conditioned on the event $E$, that there exists some bit in the buffer equal to $1$, the position of the \emph{first} such bit is a zero-indexed draw from the Geometric distribution $Geom(p)$ with $p = 0.5$ by definition of a Geometric random variable. If the event $E$ does not occur, the function returns \texttt{None}. 
+\end{proof}
+
+\subsection{Pseudocode \& proof of \texttt{sample\_bernoulli}}
+
+\begin{lstlisting}[language=Python, escapechar=|]
+# returns a single bit with some probability of success
+def sample_bernoulli(prob : T, constant_time : bool) -> bool:
+    if (prob < 0 or prob > 1): |\label{line:rangecheck}|
+        raise Exception("probability is not within [0, 1]")
+        
+            
+    if prob is 0: |\label{line:0check}|
+        return False
+        
+    if prob is 1: |\label{line:1check}|
+        return True
+        
+    # prepare for sampling first heads index by coin flipping
+    max_coin_flips = exact_int_cast(T::EXPONENT_BIAS) + exact_int_cast(T::MANTISSA_BITS) |\label{line:maxcoinflips}|
```

Comment:
A reader would wonder, "it's an exact cast to what?"

It's a cast from T::Bits (which is u32 or u64) into a usize.